src/python/grpcio_testing/grpc_testing/_channel.py

```diff
@@ -0,0 +1,753 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Channel fixtures."""
+
+import collections
+import logging
+import threading
+
+import grpc
+import grpc_testing
+from grpc_testing import _common
+
+_NOT_YET_OBSERVED = object()
+
+
+class _RpcState(object):
+
+    def __init__(self, invocation_metadata, requests, requests_closed):
+        self.condition = threading.Condition()
+        self.invocation_metadata = invocation_metadata
+        self.requests = requests
+        self.requests_closed = requests_closed
+        self.initial_metadata = None
+        self.responses = []
+        self.trailing_metadata = None
+        self.code = None
+        self.details = None
+
+
+def _state_add_request(state, request):
+    with state.condition:
+        if state.code is None and not state.requests_closed:
+            state.requests.append(request)
+            state.condition.notify_all()
+            return True
+        else:
+            return False
+
+
+def _state_no_more_requests(state):
+    with state.condition:
+        if state.code is None and not state.requests_closed:
+            state.requests_closed = True
+            state.condition.notify_all()
+
+
+def _state_take_response(state):
+    with state.condition:
+        while True:
+            if state.code is grpc.StatusCode.OK:
+                if state.responses:
+                    response = state.responses.pop(0)
+                    return _common.ChannelRpcRead(response, None, None, None)
+                else:
+                    return _common.ChannelRpcRead(None, state.trailing_metadata,
+                                                  grpc.StatusCode.OK,
+                                                  state.details)
+            elif state.code is None:
+                if state.responses:
+                    response = state.responses.pop(0)
+                    return _common.ChannelRpcRead(response, None, None, None)
+                else:
+                    state.condition.wait()
+            else:
+                return _common.ChannelRpcRead(None, state.trailing_metadata,
+                                              state.code, state.details)
+
+
+def _state_cancel(state, code, details):
+    with state.condition:
+        if state.code is None:
+            if state.initial_metadata is None:
+                state.initial_metadata = _common.FUSSED_EMPTY_METADATA
+            state.trailing_metadata = _common.FUSSED_EMPTY_METADATA
+            state.code = code
+            state.details = details
+            state.condition.notify_all()
+            return True
+        else:
+            return False
+
+
+def _state_terminate(state):
+    with state.condition:
+        while True:
```

Comment:
_(Right. And I also enjoy these kinds of explorations of ostensibly trivial concerns.)_

Let's not get too far ahead of ourselves. While [I do enjoy telling other folks how to code](https://www.youtube.com/watch?v=aOEfIrC07XA), this is only the first time I've thought through this question, so any results may be shaky and certainly won't have time and experience behind them. So I'm trying to guide myself but I'm a long way from saying others should follow.

What's attractive to me about `while True:` everywhere is that (1) it's one form that applies generally rather than one general form plus a different form in a special case and (2) it's one statement rather than two.

But it would be a huge leap to go from "`while True:` is more attractive to @nathanielmanistaatgoogle, today, for reasons he's only just recently bothered thinking about" to "`while <condition>:` is bad and should be avoided".

I agree that it's hard to imagine the logic getting _textually shorter_ than
```
while <condition>:
  <single simple statement>
return <simple expression>
```
. For some folks textual brevity _is_ simplicity, but I (for the time being, at least) prefer
```
while True:
  <parse tree node>
```
to
```
while <condition>:
  <parse tree node>
<parse tree node>
```
, even if my `<parse tree node>` isn't a _leaf_ node. I'll try to meditate on why, certainly before I start making recommendations and rules that would apply to others. It's probably got something to do with why [I prefer `for`/`else` even with loop bodies that `return` and do not `break`](https://github.com/grpc/grpc/blob/37e06cfc4d7e6473b000d4b64c7305da8e98e456/.pylintrc#L67-L69).

Certainly nothing prevents us from using `while True:` everywhere (at least as I understand the scope of this conversation). In subsequent drafts of this pull request I've tried to use `while True:` consistently throughout. (No need at this time to go mucking about the extant codebase.)